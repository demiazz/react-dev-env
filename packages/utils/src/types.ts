type JSONPrimitive = string | number | boolean | null;

type JSONObject = {
  [property: string]: JSONValue;
};

type JSONArray = JSONValue[];

export type JSONValue = JSONPrimitive | JSONObject | JSONArray;
