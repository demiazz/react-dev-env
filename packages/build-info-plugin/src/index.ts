import { Compiler, Plugin } from "webpack";
import { RawSource } from "webpack-sources";

type BuildInfoPrimitive = string | number | boolean;
type BuildInfoObject = {
  [property: string]: BuildInfo;
};
type BuildInfoArray = BuildInfo[];
type BuildInfo = BuildInfoPrimitive | BuildInfoObject | BuildInfoArray;

interface Options {
  beautify?: boolean;
  buildInfo?: BuildInfo;
  fileName?: string;
}

export class BuildInfoPlugin implements Plugin {
  private readonly beautify: boolean;
  private readonly buildInfo: BuildInfo;
  private readonly fileName: string;

  constructor({
    beautify = false,
    buildInfo = {},
    fileName = "build-info.json"
  }: Options = {}) {
    this.beautify = beautify;
    this.fileName = fileName;
    this.buildInfo = buildInfo;
  }

  apply(compiler: Compiler): void {
    compiler.hooks.emit.tap("build-info", compilation => {
      compilation.assets[this.fileName] = this.createAsset();
    });
  }

  private createAsset() {
    const json = JSON.stringify(this.buildInfo, null, this.beautify ? 2 : 0);

    return new RawSource(json);
  }
}
