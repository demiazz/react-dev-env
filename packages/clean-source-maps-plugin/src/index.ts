import { join } from "path";

import rimraf from "rimraf";
import { Compiler, Plugin } from "webpack";

export class CleanSourceMapsPlugin extends Plugin {
  apply(compiler: Compiler): void {
    compiler.hooks.done.tapAsync(
      "@rae/clean-source-maps-plugin",
      async (_, done) => {
        const outputDir = compiler.options.output?.path ?? "dist";
        const glob = join(outputDir, "**/*.@(js|css).map?(.br|.gz)");

        rimraf(glob, done);
      }
    );
  }
}
