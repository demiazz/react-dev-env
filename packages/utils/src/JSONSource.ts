import { RawSource } from "webpack-sources";

type JSONPrimitive = string | number | boolean | null;

type JSONObject = {
  [property: string]: JSONValue;
};

type JSONArray = JSONValue[];

type JSONValue = JSONPrimitive | JSONObject | JSONArray;

export class JSONSource extends RawSource {
  public constructor(value: JSONValue, beautify = false) {
    const source = JSON.stringify(value, null, beautify ? 2 : 0);

    super(source);
  }
}
