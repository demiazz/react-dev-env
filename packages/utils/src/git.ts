import { exec as execWithCallback } from "child_process";
import { promisify } from "util";

const exec = promisify(execWithCallback);

async function command(cmd: string): Promise<string> {
  const result = await exec(cmd, { cwd: __dirname });

  return result.stdout.split("\n").join("");
}

export function branch(): Promise<string> {
  return command("git rev-parse --abbrev-ref HEAD");
}

export function revision(): Promise<string> {
  return command("git rev-parse HEAD");
}
