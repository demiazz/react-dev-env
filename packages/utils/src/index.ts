export { JSONSource } from "./JSONSource";
export { JSONValue } from "./types";
