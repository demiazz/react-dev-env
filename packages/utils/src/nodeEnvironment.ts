export function setDefaultNodeEnvironment(
  defaultNodeEnvironment = "development"
) {
  const { NODE_ENV: nodeEnvironment } = process.env;

  if (nodeEnvironment == null || nodeEnvironment === "") {
    console.warn(
      'The default "NODE_ENV" environment variable was not specified. ' +
        'The "NODE_ENV" environment variable has been set to "development".'
    );

    process.env.NODE_ENV = defaultNodeEnvironment;
  }
}

export function validateNodeEnvironment(
  allowedEnvironments: string[],
  nodeEnvironment: string | undefined
) {
  if (nodeEnvironment == null) {
    throw new Error(
      'The "NODE_ENV" environment variable is required but was not specified'
    );
  }

  if (!allowedEnvironments.includes(nodeEnvironment)) {
    throw new Error(
      'Please specify a valid "NODE_ENV" environment variable. ' +
        'Valid values are "development" or "production". ' +
        `Instead, received: ${JSON.stringify(nodeEnvironment)}.`
    );
  }
}
