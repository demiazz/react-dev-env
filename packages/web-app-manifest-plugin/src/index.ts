import validateOptions from "schema-utils";
import { Compiler, Plugin } from "webpack";

import { injectToHTML } from "./html";
import { prepareAssetsAndManifest } from "./manifest";
import manifestOptionsSchema from "./schemas/manifestOptions.json";
import optionsSchema from "./schemas/options.json";
import { ManifestOptions } from "./types";
import {
  context,
  findHtmlPlugin,
  publicPath,
  writeAssets
} from "./webpack-utils";

interface Options {
  appIconFileName?: string;
  beautify?: boolean;
  favIconFileName?: string;
  manifestFileName?: string;
  screenshotFileName?: string;
}

export class WebAppManifestPlugin implements Plugin {
  private readonly manifestOptions: ManifestOptions;

  private readonly beautify: boolean;

  private readonly appIconFileName: string;
  private readonly favIconFileName: string;
  private readonly manifestFileName: string;
  private readonly screenshotFileName: string;

  public constructor(manifestOptions: ManifestOptions, options: Options = {}) {
    validateOptions(manifestOptionsSchema, manifestOptions, {
      name: "WebAppManifestPlugin",
      baseDataPath: "manifest"
    });
    validateOptions(optionsSchema, options, {
      name: "WebAppManifestPlugin",
      baseDataPath: "options"
    });

    this.manifestOptions = manifestOptions;

    this.beautify = options.beautify ?? false;

    this.appIconFileName =
      options.appIconFileName ?? "web-application/icons/icon-[hash].[ext]";
    this.favIconFileName =
      options.favIconFileName ?? "web-application/favicon-[hash].[ext]";
    this.manifestFileName =
      options.manifestFileName ?? "web-application/manifest-[hash].[ext]";
    this.screenshotFileName =
      options.screenshotFileName ??
      "web-application/screenshots/screenshot-[hash].[ext]";
  }

  public apply(compiler: Compiler): void {
    const assetsAndManifestPromise = prepareAssetsAndManifest(
      this.manifestOptions,
      {
        appIconFileName: this.appIconFileName,
        beautify: this.beautify,
        context: context(compiler),
        favIconFileName: this.favIconFileName,
        manifestFileName: this.manifestFileName,
        publicPath: publicPath(compiler),
        screenshotFileName: this.screenshotFileName
      }
    );

    compiler.hooks.make.tapPromise(
      "@react-dev-env/web-app-manifest",
      async compilation => {
        const htmlPlugin = findHtmlPlugin(compiler);

        if (!htmlPlugin) {
          return;
        }

        const { manifest } = await assetsAndManifestPromise;

        htmlPlugin
          .getHooks(compilation)
          .afterTemplateExecution.tap(
            "@react-dev-env/web-app-manifest",
            data => {
              const html = injectToHTML(data.html, manifest);

              data.html = html;
            }
          );
      }
    );

    compiler.hooks.emit.tapPromise(
      "@react-dev-env/web-app-manifest",
      async compilation => {
        const { assets } = await assetsAndManifestPromise;

        writeAssets(assets, compilation);
      }
    );
  }
}
