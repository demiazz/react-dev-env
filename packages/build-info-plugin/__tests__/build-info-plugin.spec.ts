import { readFile, rmdir } from "fs";
import { join, resolve } from "path";
import webpack, { Configuration } from "webpack";

import { BuildInfoPluginOptions, BuildInfoPlugin } from "../src";

const fixturesDir = resolve(__dirname, "fixtures");
const entryPath = join(fixturesDir, "index.js");
const buildDir = join(fixturesDir, "build");

function build(options?: BuildInfoPluginOptions): Promise<void> {
  return new Promise((resolve, reject) => {
    const configuration: Configuration = {
      context: fixturesDir,

      entry: entryPath,

      output: {
        path: buildDir
      },

      plugins: [new BuildInfoPlugin(options)]
    };

    webpack(configuration, (error, stats) => {
      if (error != null || stats.hasErrors()) {
        reject();
      }

      resolve();
    });
  });
}

function readBuildInfo(options?: BuildInfoPluginOptions): Promise<string> {
  return new Promise((resolve, reject) => {
    const buildInfoPath = join(
      buildDir,
      options?.fileName || "build-info.json"
    );

    readFile(
      buildInfoPath,
      {
        encoding: "utf8"
      },
      (err, data) => {
        if (err) {
          return reject(err);
        }

        resolve(data);
      }
    );
  });
}

async function buildAndMatch(options?: BuildInfoPluginOptions) {
  await build(options);

  expect(await readBuildInfo(options)).toMatchSnapshot();
}

describe("BuildInfoPlugin", () => {
  afterEach(done => {
    rmdir(buildDir, { recursive: true }, done);
  });

  describe("default options", () => {
    it('creates a "build-info.json" file with empty JSON', async () => {
      await buildAndMatch();
    });
  });

  describe("with user's build info", () => {
    it('creates a "build-info.json" file with given build info', async () => {
      await buildAndMatch({
        buildInfo: {
          release: "1.2.3"
        }
      });
    });
  });

  describe('with enabled "beautify" option', () => {
    it('creates a "build-info.json" file with formatted build info', async () => {
      await buildAndMatch({
        beautify: true,
        buildInfo: {
          release: "2.3.4"
        }
      });
    });
  });

  describe("with custom file name", () => {
    it("creates file with given build info and given file name", async () => {
      await buildAndMatch({
        buildInfo: {
          release: "2.3.4"
        },
        fileName: "static/release.json"
      });
    });
  });
});
