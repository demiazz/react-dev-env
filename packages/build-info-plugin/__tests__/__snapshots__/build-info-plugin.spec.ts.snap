// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuildInfoPlugin default options creates a "build-info.json" file with empty JSON 1`] = `"{}"`;

exports[`BuildInfoPlugin with custom file name creates file with given build info and given file name 1`] = `"{\\"release\\":\\"2.3.4\\"}"`;

exports[`BuildInfoPlugin with enabled "beautify" option creates a "build-info.json" file with formatted build info 1`] = `
"{
  \\"release\\": \\"2.3.4\\"
}"
`;

exports[`BuildInfoPlugin with user's build info creates a "build-info.json" file with given build info 1`] = `"{\\"release\\":\\"1.2.3\\"}"`;
